// @kushki/js/lib/Kushki.js
"use strict";
exports.__esModule = true;
exports.Kushki = void 0;
var Identifiers_1 = require("./../lib/constant/Identifiers");
var KushkiError_1 = require("./../lib/generic/KushkiError");
var Container_1 = require("./../lib/infrastructure/Container");
require("reflect-metadata");
var UtilsService_1 = require("./../lib/service/UtilsService");
var KushkiInfo_1 = require("./../lib/KushkiInfo");
/**
 * Implementation
 */
var Kushki = /** @class */ (function () {
    // tslint:disable-next-line:max-func-body-length
    function Kushki(body) {
        KushkiInfo_1.KInfo.sKushkiInfo = KushkiInfo_1.KInfo.buildKushkiInfo(body.kushkiInfo);
        body.inTestEnvironment =
            body.inTestEnvironment === undefined ? false : body.inTestEnvironment;
        body.regional = body.regional === undefined ? false : body.regional;
        this._merchantId = body.merchantId;
        this._inTestEnvironment = body.inTestEnvironment;
        this._regional = body.regional;
        this._cardService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CardService)(this._regional);
        this._siftScience = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.SiftScienceService)(this._regional);
        this._transferService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.TransferService)(this._regional);
        this._cashService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CashService)(this._regional);
        this._multiMerchantInfoService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.MultiMerchantService)(this._regional);
        this._commissionService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CommissionService)(this._regional);
        this._kushkiService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.KushkiService)(this._regional);
        this._secureService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.SecureService)();
        this._cardAsyncService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CardAsyncService)();
        this._payoutsCashService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.PayoutsCashService)(this._regional);
        this._payoutsTransferService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.PayoutsTransferService)(this._regional);
        this._transferSubscriptionService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.TransferSubscriptionService)(this._regional);
        this._mobileProcessorService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.MobileProcessorService)();
        this._authService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.AuthService)();
        this._kPayService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.KPayService)();
        this._cardDynamicService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CardDynamicService)();
        this._cardSubscriptionDynamicService = Container_1.CONTAINER.get(Identifiers_1.IDENTIFIERS.CardSubscriptionDynamicService)();
    }
    Kushki.prototype.requestToken = function (body, callback) {
        this._cardService.requestToken(body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype.requestInitAntiFraud = function (userId, callback) {
        this._resolve(this._siftScience.requestInitAntiFraud(userId, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    /**
     * @deprecated use requestDeviceToken instead
     * @param {Function} body with value
     * @param {Function} callback with value
     * @returns {void}
     */
    Kushki.prototype.requestTokenCharge = function (body, callback) {
        this._requestDeviceToken(body, callback);
    };
    Kushki.prototype.requestDeviceToken = function (body, callback) {
        this._requestDeviceToken(body, callback);
    };
    Kushki.prototype.requestSubscriptionToken = function (body, callback) {
        this._cardService.requestSubscriptionToken(body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype.requestTransferToken = function (body, callback) {
        this._resolve(this._transferService.requestTransferToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.checkStatus = function (callback) {
        this._resolve(this._kushkiService.checkStatus(this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    /**
     * @deprecated since version 1.19.8 use requestBankList instead
     * @param {Function} callback with value
     * @returns {void}
     */
    Kushki.prototype.requestPseBankList = function (callback) {
        this._requestBankList(callback);
    };
    Kushki.prototype.requestBankList = function (callback) {
        this._requestBankList(callback);
    };
    Kushki.prototype.requestDeferred = function (body, callback) {
        this._resolve(this._cardService.requestDeferred(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestBinInfo = function (body, callback) {
        this._cardService
            .requestBinInfo(body, this._merchantId, this._inTestEnvironment, this._regional)
            .subscribe({
            error: function (err) {
                return Kushki.callbackError(err, callback);
            },
            next: function (response) {
                callback(response);
            }
        });
    };
    Kushki.prototype.requestBrandsByMerchant = function (callback) {
        this._kushkiService
            .requestBrandsByMerchant(this._merchantId, this._inTestEnvironment, this._regional)
            .subscribe({
            error: function (err) {
                return Kushki.callbackError(err, callback);
            },
            next: function (response) {
                callback(response);
            }
        });
    };
    Kushki.prototype.requestBrandsLogosByMerchant = function (callback) {
        this._kushkiService
            .requestBrandsLogosByMerchant(this._merchantId, this._inTestEnvironment, this._regional)
            .subscribe({
            error: function (err) {
                return Kushki.callbackError(err, callback);
            },
            next: function (response) {
                callback(response);
            }
        });
    };
    Kushki.prototype.requestCashToken = function (body, callback) {
        this._resolve(this._cashService.requestCashToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestSecureServiceValidation = function (body, callback) {
        this._resolve(this._secureService.requestSecureServiceValidation(this._merchantId, body, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestCardAsyncToken = function (body, callback) {
        this._resolve(this._cardAsyncService.requestCardAsyncToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestSubscriptionCardAsyncToken = function (body, callback) {
        this._resolve(this._cardAsyncService.requestCardSubscriptionAsyncToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.multiMerchantInfo = function (body, callback) {
        this._resolve(this._multiMerchantInfoService.multiMerchantInfo(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestPayoutsCashToken = function (body, callback) {
        this._resolve(this._payoutsCashService.requestPayoutsCashToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestPayoutsTransferToken = function (body, callback) {
        this._resolve(this._payoutsTransferService.requestPayoutsTransferToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.getCommissionConfiguration = function (body, callback) {
        this._resolve(this._commissionService.getCommissionConfiguration(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestTransferSubscriptionToken = function (body, callback) {
        this._resolve(this._transferSubscriptionService.requestToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestTransferSubscriptionBankList = function (callback) {
        this._resolve(this._transferSubscriptionService.requestBankList(this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestPayoutsTransferBankList = function (callback) {
        this._resolve(this._payoutsTransferService.requestPayoutsTransferBankList(this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestMobileProcessorToken = function (body, callback) {
        this._resolve(this._mobileProcessorService.requestToken(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestCardDynamicToken = function (bin, body, callback) {
        this._cardDynamicService.requestCardDynamicToken(bin, body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype.requestCardSubscriptionDynamicToken = function (bin, body, callback) {
        this._cardSubscriptionDynamicService.requestCardSubscriptionDynamicToken(bin, body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype.requestInitAuth = function (body, callback) {
        this._resolve(this._authService.requestInitAuth(body, this._inTestEnvironment), callback);
    };
    Kushki.prototype.deletePaymentMethod = function (body, callback) {
        this._resolve(this._kPayService.deletePaymentMethod(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestSavedPaymentMethods = function (body, callback) {
        this._resolve(this._kPayService.requestSavedPaymentMethods(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.requestVerifyAuth = function (body, callback) {
        this._resolve(this._authService.requestVerifyAuth(body.code, this._inTestEnvironment), callback);
    };
    Kushki.prototype.validateSession = function (callback) {
        this._resolve(this._authService.validateSession(this._inTestEnvironment), callback);
    };
    Kushki.prototype.requestValidate3DS = function (body, callback) {
        this._cardService.validate3DS(body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype.requestSecureInit = function (body, callback) {
        this._resolve(this._cardService.requestSecureInit(body, this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype.initVisaBrandingAnimation = function (callback, body) {
        this._kushkiService.initVisaBrandingAnimation(this._inTestEnvironment, callback, body);
    };
    Kushki.callbackError = function (err, callback) {
        if (err instanceof KushkiError_1.KushkiError)
            callback({
                code: "" + err.code,
                error: err.code + " - " + err.getMessage(),
                message: "" + err.getMessage()
            });
        else
            callback({
                code: "" + UtilsService_1.UtilsService.sGet(err, "response.data.code", "K002"),
                error: "" + UtilsService_1.UtilsService.sGet(err, "message", "Ha ocurrido un error inesperado."),
                message: "" + UtilsService_1.UtilsService.sGet(err, "response.data.message", "Ha ocurrido un error inesperado.")
            });
    };
    Kushki.prototype._requestBankList = function (callback) {
        this._resolve(this._transferService.requestBankList(this._merchantId, this._inTestEnvironment, this._regional), callback);
    };
    Kushki.prototype._requestDeviceToken = function (body, callback) {
        this._cardService.requestDeviceToken(body, this._merchantId, this._inTestEnvironment, this._regional, callback);
    };
    Kushki.prototype._resolve = function (observable, callback) {
        observable.subscribe({
            error: function (err) { return Kushki.callbackError(err, callback); },
            next: function (response) {
                callback(response);
            }
        });
    };
    return Kushki;
}());
exports.Kushki = Kushki;
