// @kushki/js/lib/constant/Identifiers.js
"use strict";
/**
 * Injection identifiers
 */
exports.__esModule = true;
exports.IDENTIFIERS = void 0;
var IDENTIFIERS = {
    AntiFraud: Symbol["for"]("AntiFraud"),
    AuthService: Symbol["for"]("AuthService"),
    CardAsyncService: Symbol["for"]("CardAsyncService"),
    CardDynamicService: Symbol["for"]("CardDynamicService"),
    CardService: Symbol["for"]("CardService"),
    CardSubscriptionDynamicService: Symbol["for"]("CardSubscriptionDynamicService"),
    CashService: Symbol["for"]("CashService"),
    CommissionService: Symbol["for"]("CommissionService"),
    KPayService: Symbol["for"]("KPayService"),
    KushkiGateway: Symbol["for"]("KushkiGateway"),
    KushkiService: Symbol["for"]("KushkiService"),
    MobileProcessorService: Symbol["for"]("MobileProcessorService"),
    MultiMerchantService: Symbol["for"]("MultiMerchantService"),
    PayoutsCashService: Symbol["for"]("PayoutsCashService"),
    PayoutsTransferService: Symbol["for"]("PayoutsTransferService"),
    SecureService: Symbol["for"]("SecureService"),
    SiftScienceService: Symbol["for"]("SiftScienceService"),
    TransferService: Symbol["for"]("TransferService"),
    TransferSubscriptionService: Symbol["for"]("TransferSubscriptionsService")
};
exports.IDENTIFIERS = IDENTIFIERS;
